sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala

```diff
@@ -318,10 +318,31 @@ private[parquet] class ParquetRowConverter(
         new ParquetMapConverter(parquetType.asGroupType(), t, updater)
 
       case t: StructType =>
+        val wrappedUpdater = {
+          if (updater.isInstanceOf[RowUpdater]) {
+            // `updater` is a RowUpdater, implying that the parent container is a struct.
+            // We do NOT need to perform defensive copying here because either:
+            //
+            //   1. The path from the schema root to this field consists only of nested
+            //      structs, so this converter will only be invoked once per record and
+            //      we don't need to copy because copying will be done in the final
+            //      UnsafeProjection, or
+            //   2. The path from the schema root to this field contains a map or array,
+            //      in which case we will perform a recursive defensive copy via the
```

Comment:
Correctness relies on the copy actually being a deep copy. Looking elsewhere in this file, we have comments like

```
    // NOTE: We can't reuse the mutable Map here and must instantiate a new `Map` for the next
    // value.  `Row.copy()` only copies row cells, it doesn't do deep copy to objects stored in row
    // cells.
```

which suggest that certain copying might be shallow, so it's important to double-check and make sure that the copies are indeed deep.

Here, the state being copied is an `InternalRow`. To be more specific, it's actually a `SpecificInternalRow` (I'll update the `.asInstanceOf` cast below to reflect this). `SpecificInternalRow` extends `BaseGenericInternalRow` and #18483 changed that to implement a deep-copy, recursively copying maps, arrays, and structs.

Reply:
I think the existing comment about `Row.copy()` is outdated, so we might be able to optimize those other parts of the code, too; I'm going to defer that to future work / another PR, though.

Reply:
**Update**: in #27089 I'm removing these other unnecessary ArrayBuffer copies.